python/ray/dag/dag_node_operation.py

```diff
@@ -0,0 +1,102 @@
+    1. READ: Read from an input channel.
+    2. COMPUTE: Execute the method corresponding to the node.
+    3. WRITE: Write to an output channel.
+    """
+
+    READ = "READ"
+    COMPUTE = "COMPUTE"
+    WRITE = "WRITE"
+
+
+@DeveloperAPI
+class DAGNodeOperation:
+    def __init__(
+        self,
+        idx: int,
+        operation_type: DAGNodeOperationType,
+    ):
+        """
+        Args:
+            idx: The index of the task that this operation belongs to
+                in the actor's ExecutableTask list. The index is not
+                the same as bind_index, but there are positive correlations
+                between the two.
+            operation_type: The type of operation to perform.
+        """
+        self.idx = idx
+        self.type = operation_type
+
+
+@DeveloperAPI
+@total_ordering
+class DAGOperationGraphNode:
+    def __init__(
+        self,
+        operation: DAGNodeOperation,
+        idx: int,
+        actor_handle: "ray.actor.ActorHandle",
+        requires_nccl: bool,
+    ):
+        """
+        DAGOperationGraphNode represents a node in the DAG operation graph.
+        It contains information about the node's in-degree, out-degree, edges,
+        and the operation it performs.
+
+        Args:
+            operation: The operation that this node performs. The operation
+                can be a READ, COMPUTE, or WRITE operation.
+            idx: A unique index into the original DAG.
+            dag_node: The DAGNode that this operation belongs to.
+        """
+        self.operation = operation
+        self.idx = idx
+        self.actor_handle = actor_handle
+        self.requires_nccl = requires_nccl
+        self.in_edges = set()
+        self.out_edges = set()
+
+    @property
+    def in_degree(self) -> int:
+        return len(self.in_edges)
+
+    def __lt__(self, other):
+        """
+        Two DAGOperationGraphNodes are comparable only when they belong to
+        the same actor. For operations on the same actor, if idx is smaller,
+        the DAGNode to which this operation belongs has a smaller `bind_index`.
+        """
+        assert self.actor_handle == other.actor_handle
+        return self.operation.idx < other.operation.idx
+
+    def __eq__(self, other):
+        """
+        Two DAGOperationGraphNodes are comparable only when they belong to the
+        same actor. For operations on the same actor, two operations are equal
+        only when they have the same `idx` and `type`.
+        """
+        assert self.actor_handle == other.actor_handle
+        return (
+            self.operation.idx == other.operation.idx
+            and self.operation.type == other.operation.type
+        )
+
+    def __hash__(self):
+        return hash((self.operation, self.idx))
+
+    def add_edge(self, out_node: "DAGOperationGraphNode"):
+        """
+        Add an edge from this node to `out_node`. An edge is a tuple of
+        the operation's index and type.
```

Comment:
nit; maybe it is better we just make util funtion?
```
def add_edge(from: DAGOperationGraphNode, to: DAGOperationGraphNode):
    ...
```

This way, we can avoid we change the state of the input argument (which is usually not a good pattern)

Reply:
> we can avoid we change the state of the input argument

I can't get the point. We still need to update `from.in_edges` and `to.out_edges`. Would you mind explaining more?

Reply:
I feel a bit ugly DagNode modifies the state of other DagNode implicitly. (vs function is clear it just modifies the state of given 2 arguments). but it is nit. I will leave it up to you.

Reply:
Got your point. I will update it

Reply:
Updated https://github.com/ray-project/ray/pull/46911/commits/e0df08bf80446f0f238a1926ac3add7d58916d1d